Add frontend for creating PipelineResource

Megan-Wright opened the pull request:
For https://github.com/tektoncd/dashboard/issues/247

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

- Adds ability to create a PipelineResource through the dashboard

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

Megan-Wright commented:
/uncc @CarolynMabbott @vtereso 
/cc @a-roberts @AlanGreene

a-roberts commented on src/containers/PipelineResourcesModal/PipelineResourcesModal.js:
```diff
@@ -0,0 +1,334 @@
+import GitResourceFields from '../../components/PipelineResourcesModal/GitResourceFields';
+import UniversalFields from '../../components/PipelineResourcesModal/UniversalFields';
+// import BasicAuthFields from '../../components/SecretsModal/BasicAuthFields';
+// import { createSecret } from '../../actions/secrets';
```
👀

Megan-Wright replied:
oops, cleared that up 😸

a-roberts commented on src/containers/PipelineResourcesModal/PipelineResourcesModalErrorNotification.test.js:
```diff
@@ -0,0 +1,93 @@
+    default: [
+      {
+        uid: '0',
+        name: 'github-repo-access-secret',
```
Isn't this actually a PipelineResource name?

Megan-Wright replied:
ahh yep - fixed now 😸

a-roberts commented on src/components/PipelineResourcesModal/UniversalFields.js:
```diff
@@ -0,0 +1,84 @@
+        labelText="Name"
+        onChange={handleChangeTextInput}
+        invalid={'name' in invalidFields}
+        invalidText="Must be less than 64 characters and contain only lowercase alphanumeric characters or -"
```
Why 64?

```
(base) adams-mbp:devops-world-example aroberts$ k apply -f resource.yaml 
pipelineresource.tekton.dev/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa created
(base) adams-mbp:devops-world-example aroberts$ k get pipelineresource
NAME                                                                                                                                                           AGE
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa   10s
```

```
: Invalid value: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa": must be no more than 253 characters
```

Megan-Wright replied:
Tested it with postman - 63 is the max number of characters the name allows for with our API

a-roberts replied:
Fair enough

a-roberts commented on src/components/PipelineResourcesModal/UniversalFields.js:
```diff
@@ -0,0 +1,84 @@
+        onChange={handleChangeTextInput}
+        invalid={'url' in invalidFields}
+        invalidText="URL required"
+        autoComplete="off"
```
Why off?

Megan-Wright replied:
idk, just copied it from the universal fields for secrets which all had auto-complete off

a-roberts commented:
@AlanGreene can we get a look as well please? 😸

CarolynMabbott commented:
Creating a Image Pipeline Resource doesnt seem to work on your PR, have tried with:
cat
tekton-pipelines
Image
cat goes meow

a-roberts commented:
> As I mentioned on the [issue](https://github.com/tektoncd/dashboard/issues/247#issuecomment-504084508) I really would have liked to see some design input before work started on this. There are a lot of resource types (currently: git, pullRequest, image, cluster, storage (gcs, build-gcs), cloudevent) that have slightly different configs that need to be accounted for. This PR as it currently stands only supports git and image.
> 
> Is there a particular requirement driving this right now? Do we actually need/want a UI to create PipelineResources, or should we be encouraging a pipeline-as-code approach where possible?
> 
> Given the recent design direction to move the Create Secret and Create PipelineRun UIs to full page experience instead of modals, we should follow that approach if we're proceeding with this new UI, rather than introducing something we know we're going to have to change in the very near future anyway.

Yeah, that's a good point, there isn't a particular requirement driving this currently - I think it was a case of "we can let users create PipelineRuns and secrets, logically one would want to create PipelineResources too as that really completes the picture in terms of kicking off something meaningful using just the UI"), but you're right there are plenty of other types we would want to add.

@Megan-Wright do you think it'd be particularly difficult to add other resources types, or at least make it easy for anybody else to do so? 

We could always leverage Megan's work as part of a full page create resource experience either way so the code here should definitely be handy regardless of our decision - this just happened to be a relatively high priority item on the ZenHub board and hence the sudden progress on this one.

Megan-Wright commented:
@a-roberts Yeah it should be fairly easy to add other types of resource - though I've never seen any of the types that @AlanGreene mentioned so would need some examples to work off, but providing it's just changing the `params` array it should be doable.

Megan-Wright commented:
> Creating a Image Pipeline Resource doesnt seem to work on your PR, have tried with:
> cat
> tekton-pipelines
> Image
> cat goes meow

Fixed now - turns out I deleted a line of code somewhere by accident that was making this fail......

Megan-Wright commented:
/retest

a-roberts commented:
Had a ponder about this, and we also had to debug a nice error from the proxy involving getting the bytes and converting them to utf-8 - Megan was missing `Kind` for image and we catch proxy errors and just return a 404.

I think it's good to add this PR as is for several reasons.

- Git and Image types are what we think are the most common resources in Tekton (if we assume CI/CD is the basic flow)
- Yeah, we do want to move to full-screen instead of modal pages for PipelineRuns and Secrets eventually, that's certainly a different piece of work though and by having this in, we'll at least have something for design that's more: we have these three modes you can see and use right now and we want them to be big full screen ones
- There's no harm in incrementally improving, we could quite easily make additional issues for
Cluster, PullRequest, Storage and CloudEvent

I think that these four are going to be non-trivial to test out and worthy of doing as their own piece.

- This PR's already XXL 😄 

@AlanGreene what do you think?

AlanGreene commented on src/components/PipelineResourcesModal/GitResourceFields.js:
```diff
@@ -0,0 +1,34 @@
+      id="revision"
+      placeholder="pipeline-resource-revision"
+      value={revision}
+      labelText="Revision"
```
we need to use react-intl for user-facing strings

same for UniversalFields.js and PipelineResource.js

Megan-Wright replied:
ok, fixed this now :)

AlanGreene replied:
still seeing hard-coded string here `labelText="Revision"`

two more `labelText` props below:
- https://github.com/tektoncd/dashboard/pull/942/files#diff-58230c181d9089fb7a4c65e207282174R43
- https://github.com/tektoncd/dashboard/pull/942/files#diff-58230c181d9089fb7a4c65e207282174R89

Megan-Wright replied:
oops, changed that now

Megan-Wright commented:
@AlanGreene I fixed up everything concerning your review - and all tests are passing now... are we good to get this in?

Megan-Wright commented:
/retest

AlanGreene commented:
This looks good for now, thanks @Megan-Wright 
/lgtm
